src/lsptoolshost/services/projectContextService.ts

```diff
@@ -0,0 +1,43 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { RoslynLanguageServer } from '../roslynLanguageServer';
+import { VSGetProjectContextsRequest, VSProjectContext, VSProjectContextList } from '../roslynProtocol';
+import { TextDocumentIdentifier } from 'vscode-languageserver-protocol';
+import { UriConverter } from '../uriConverter';
+
+export class ProjectContextService {
+    /** Track the project contexts for a particular document uri. */
+    private _projectContexts: { [uri: string]: Promise<VSProjectContextList> | VSProjectContextList } = {};
+
+    constructor(private languageServer: RoslynLanguageServer) {}
+
+    clear() {
+        this._projectContexts = {};
+    }
+
+    async getCurrentProjectContext(uri: string | vscode.Uri): Promise<VSProjectContext | undefined> {
+        const projectContexts = await this.getProjectContexts(uri);
+        return projectContexts?._vs_projectContexts[projectContexts._vs_defaultIndex];
+    }
+
+    async getProjectContexts(uri: string | vscode.Uri): Promise<VSProjectContextList | undefined> {
+        const uriString = uri instanceof vscode.Uri ? UriConverter.serialize(uri) : uri;
+
+        if (!(uriString in this._projectContexts)) {
```

Comment:
A little concerned about this caching.  It kind of seems like we should just query the server every time the active text editor changes, regardless of if we had contexts before.  It isn't a hugely expensive request.

For example, if we removed/added a TFM, this cache doesn't seem like it would get refreshed.

I believe VS already even sends the request on typing.

Reply:
Removing caching from this PR